TykTechnologies/tyk-identity-broker
# Support for mongo db as DataLoader

**What was done?**

- Implemented the methods of DataLoader interface for Mongo, so, now we can use this database to persist the data related to the profiles. The methods are: Init, LoadIntoStore and Flush

- Reestructured code of DataLoader, now its a separated package

- Moved configurations to its own package

**Breaking changes**
None

**New block in configuration file**
A block was added to detail the data correspondent for Mongo as DataStorage:

```
"Storage":{
        "storage_type":"mongo",
        "mongo":{
            "mongo_url": "mongodb://tyk-mongo:27017/tyk_tib"
        }
    }
```

The definition for each field is as next:

- storage_type: which storage type you want use, the valid values are: mongo | file. If the block is not present in the config file then File is assumed as default
- mongo.mongo_url: url to connect to the database

**How to test it?**

Just execute TIB as usually you do, but provide a configuration file that contains the block described before, in this case is not needed to pass a profile file as parameter. In the logs you will see that the DataLoader with Mongo was initiated and it loaded the profilles from mongo. To test using Curl do the next steps:

- **Adding a Profile**
```
curl -X POST \
  http://localhost:3010/api/profiles/1 \
  -H 'Accept: */*' \
  -H 'Accept-Encoding: gzip, deflate' \
  -H 'Authorization: test-secret' \
  -H 'Cache-Control: no-cache' \
  -H 'Connection: keep-alive' \
  -H 'Content-Length: 792' \
  -H 'Content-Type: text/plain' \
  -H 'Host: localhost:3010' \
  -H 'Postman-Token: 22f8732e-f329-4605-a6f8-236167116ee1,48f13a84-068b-46e0-aa2a-205ac92eca70' \
  -H 'User-Agent: PostmanRuntime/7.20.1' \
  -H 'cache-control: no-cache' \
  -d '{
			"ActionType": "GenerateTemporaryAuthToken",
			"ID": "1",
			"IdentityHandlerConfig": {
				"DashboardCredential": "822f2b1c75dc4a4a522944caa757976a",
				"DisableOneTokenPerAPI": false,
				"TokenAuth": {
					"BaseAPIID": "e1d21f942ec746ed416ab97fe1bf07e8"
				}
			},
			"MatchedPolicyID": "5654566b30c55e3904000003",
			"OrgID": "53ac07777cbb8c2d53000002",
			"ProviderConfig": {
				"ExrtactUserNameFromBasicAuthHeader": true,
				"OKCode": 200,
				"OKRegex": "origin",
				"OKResponse": "ewogICJvcmlnaW4iOiAiNjIuMjMyLjExNC4yNTAsIDE3OC42Mi4xMS42MiwgMTc4LjYyLjExLjYyIgp9Cg==",
				"TargetHost": "http://sharrow.tyk.io/ba-1/"
			},
			"ProviderConstraints": {
				"Domain": "",
				"Group": ""
			},
			"ProviderName": "ProxyProvider",
			"ReturnURL": "",
			"Type": "passthrough"
}
```
- **Saving in Mongo**
Is the equivalent for flushing changes, it will create as well a backup with the timestamp as name of collection.

```
curl -X POST \
  http://localhost:3010/api/profiles/save \
  -H 'Authorization: test-secret' \
  -H 'Postman-Token: f86621b3-8c61-4b14-ad2a-5a6c79d371c4' \
  -H 'cache-control: no-cache'
```

After executing these 2 steps you should be able to check against the database that the record was stored and also it's loaded in the selected backend (redis or in-memory)

**Notes:**

- In this PR was started the modularization of TIB in more granular packages that will allow later the integration in tyk-dashboard in a more friendly way.
- Is needed to write more tests.
